docs/type_inference.md

```diff
@@ -26,12 +26,21 @@ Both
 - **verifiers**: implemented by `Op::verify()`, and
 - **shape functions**: implemented by `InferTypeOpInterface` like `Op::inferReturnTypes()` or `Op::inferReturnTypeComponents`
 
-may have verification code to check operands/attributes/results. An ideal split would be that: let the verifiers check the operands/attributes, then let shape functions to calculate inferred result types and check the compatibility against the real result types. However, in reality this split has a few problems:
+may have verification code to check operands/attributes/results. An initial split would be that: let the verifiers check the operands/attributes, then let shape functions only calculate inferred result types and check the compatibility against the real result types. However, in reality this split has a few problems:
 
-1. Duplicated code: for example in verifiers we do some processing on the operands then verify some intermediate results, then in shape functions these intermediate results are useful to infer the final results. These intermediate results have to be calculated twice.
-2. Maintenance burden: as verifications of an op are contained in two different methods.
+- The shape function can be called by the auto-gen `builder()` alone, without calling the verifier first. So the related inputs must be verified in shape function as well.
+- Duplicated code: for example in verifiers we do some processing on the operands then verify some intermediate results, then in shape functions these intermediate results are useful to infer the final results. These intermediate results have to be calculated twice.
+- Maintenance burden: as verifications of an op are contained in two different methods.
+
+The solution is as follows:
+
+1. **For most ops w/o regions** (like PadOp):
+Put all the verification code into the shape functions, and discard verifiers totally.
+
+2. **For ops w/ regions** (like `ReduceOp/IfOp`, a full list is [here](https://github.com/openxla/stablehlo/pull/401)): the `builder()` of the op can be created before all the components are in place, for [example](https://github.com/tensorflow/mlir-hlo/blob/master/mhlo/transforms/mhlo_canonicalize_reduction/mhlo_canonicalize_reduction.cc#L221), the `ReduceOp` is created without regions and soon the shape functions are used.
+    1. If the regions are not needed for type inference (like `ReduceOp`), put the region related verification logic in verifiers. Duplicate some code if it is inevitable.
```

Comment:
why not suggest use of verifyRegions so that verification that is part of type inference doesn't need to happen again? by the time verifyRegions is invoked, type inference would have been done already. This should help avoid most of the duplication.

Reply:
Hi Smit, thanks for the suggestions: `verifyRegions()` seems not used in StableHLO / MHLO yet, but it's a good option. For the `ReduceOp`, we could consider put the region-related verification (link below)  into a new `verifyRegions()` method, but as you may notice, it relies on a lot of intermediate result from other parts of code above it. It it not clear enough for me for the moment to split them further.

As the crash that https://github.com/openxla/stablehlo/pull/760 and this PR trying to address is an P0 issue, I will submit after get approval. But we may continue the discussion in a new bug if necessary. 

region-related verification in `ReduceOp`: https://github.com/tensorflow/mlir-hlo/blame/master/stablehlo/stablehlo/dialect/TypeInference.cpp#L1551

Reply:
Mark as resolved for now, and still open to discussion in the future.